src/honey_badger.rs

```diff
@@ -200,43 +222,213 @@ where
+        let shares = &self.received_shares[&self.epoch][proposer_id];
+        if shares.len() > self.netinfo.num_faulty() {
+            if let Some(ciphertexts) = self.ciphertexts.get(&self.epoch) {
+                if let Some(ciphertext) = ciphertexts.get(proposer_id) {
+                    let ids_u64: BTreeMap<&NodeUid, u64> = shares
+                        .keys()
+                        .map(|id| (id, *self.netinfo.node_index(id).unwrap() as u64))
+                        .collect();
+                    let indexed_shares: BTreeMap<&u64, _> = shares
+                        .into_iter()
+                        .map(|(id, share)| (&ids_u64[id], share))
+                        .collect();
+                    if let Ok(decrypted_selection) = self
+                        .netinfo
+                        .public_key_set()
+                        .decrypt(indexed_shares, ciphertext)
+                    {
+                        self.decrypted_selections
+                            .insert(proposer_id.clone(), decrypted_selection);
+                        return true;
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    fn send_decryption_shares(
+        &mut self,
+        cs_output: BTreeMap<NodeUid, Vec<u8>>,
+    ) -> Result<(), Error> {
+        for (proposer_id, v) in cs_output {
+            let mut ciphertext: Ciphertext<Bls12>;
+            if let Ok(ct) = bincode::deserialize(&v) {
+                ciphertext = ct;
+            } else {
+                warn!("Invalid ciphertext from proposer {:?} ignored", proposer_id);
+                // TODO: Log the incorrect node `j`.
+                continue;
+            }
+            if let Some(share) = self.netinfo.secret_key().decrypt_share(&ciphertext) {
+                // Send the share to remote nodes.
+                self.messages.0.push_back(
+                    Target::All.message(
+                        MessageContent::DecryptionShare {
+                            proposer_id: proposer_id.clone(),
+                            share: share.clone(),
+                        }.with_epoch(self.epoch),
+                    ),
+                );
+                let our_id = self.netinfo.our_uid().clone();
+                let epoch = self.epoch;
+                // Receive the share locally.
+                self.handle_decryption_share_message(&our_id, epoch, &proposer_id, &share)?;
+            } else {
+                warn!("Share decryption failed for proposer {:?}", proposer_id);
+                // TODO: Log the decryption failure.
+                continue;
+            }
+            let ciphertexts = match self.ciphertexts.entry(self.epoch) {
+                Entry::Occupied(entry) => entry.into_mut(),
+                Entry::Vacant(entry) => entry.insert(BTreeMap::new()),
+            };
+            ciphertexts.insert(proposer_id, ciphertext);
```

Comment:
Isn't it sufficient to just store the proposer ID, without the ciphertext?

Reply:
I think not? A ciphertext is required for `PublicKeySet::decrypt` in `HoneyBadger::try_decrypt_proposer_selection`.